uvtools/dspec.py

```diff
@@ -44,67 +53,157 @@ def calc_width(filter_size, real_delta, nsamples):
+
+    # 1D clean
+    if not clean2d:
+        # setup _d and _w arrays
+        if edgecut > 0:
+            _window = np.zeros(data.shape[-1], dtype=np.float)
+            _window[edgecut:-edgecut] = gen_taper(window, data.shape[-1] - edgecut * 2, alpha=alpha)
+        else:
+            _window = gen_taper(window, data.shape[-1], alpha=alpha)
+        if dndim == 2:
+            _window = _window[None, :]
+        _d = np.fft.ifft(data * wgts * _window, axis=-1)
+        _w = np.fft.ifft(wgts * _window, axis=-1)
+
+        # calculate area array
+        area = np.ones(data.shape[-1], dtype=np.int)
+        uthresh, lthresh = calc_width(filter_size, real_delta, data.shape[-1])
+        area[uthresh:lthresh] = 0
+
+        # run clean
+        if dndim == 1:
+            # For 1D data array run once
+            _d_cl, info = aipy.deconv.clean(_d, _w, area=area, tol=tol, stop_if_div=False, maxiter=maxiter, gain=gain)
+            d_mdl = np.fft.fft(_d_cl)
+            del info['res']
+        elif data.ndim == 2:
+            # For 2D data array, iterate
+            info = []
+            d_mdl = np.empty_like(data)
+            for i in range(data.shape[0]):
+                if _w[i, 0] < skip_wgt:
+                    d_mdl[i] = 0 # skip highly flagged (slow) integrations
+                    info.append({'skipped': True})
+                else:
+                    _d_cl, info_here = aipy.deconv.clean(_d[i], _w[i], area=area, tol=tol, stop_if_div=False, maxiter=maxiter, gain=gain)
+                    d_mdl[i] = np.fft.fft(_d_cl)
+                    del info_here['res']
+                    info.append(info_here)
+
+    # 2D clean on 2D data
+    else:
+        # setup _d and _w arrays
+        if edgecut[0] > 0:
+            _w1 = np.zeros(data.shape[0], dtype=np.float)
+            _w1[edgecut[0]:-edgecut[0]] = gen_taper(window[0], data.shape[0] - edgecut[0] * 2, alpha=alpha)
+        else:
+            _w1 = gen_taper(window[0], data.shape[0], alpha=alpha)
+        if edgecut[1] > 0:
+            _w2 = np.zeros(data.shape[1], dtype=np.float)
+            _w2[edgecut[1]:-edgecut[1]] = gen_taper(window[1], data.shape[1] - edgecut[1] * 2, alpha=alpha)
+        else:
+            _w2 = gen_taper(window[1], data.shape[1], alpha=alpha)
+        _window = _w1[:, None] * _w2[None, :]
+        _d = np.fft.ifft2(data * wgts * _window, axes=(0, 1))
+        _w = np.fft.ifft2(wgts * _window, axes=(0, 1))
+
+        # calculate area array
+        a1 = np.ones(data.shape[0], dtype=np.int)
+        uthresh, lthresh = calc_width(filter_size[0], real_delta[0], data.shape[0])
+        a1[uthresh:lthresh] = 0
+        a2 = np.ones(data.shape[1], dtype=np.int)
+        uthresh, lthresh = calc_width(filter_size[1], real_delta[1], data.shape[1])
+        a2[uthresh:lthresh] = 0
+        area = np.outer(a1, a2)
+
+        # check for filt2d_mode
+        if filt2d_mode == 'plus':
```

Comment:
I'm not sure if this is exactly the same as what we're doing in smooth_cal. Also, I'm a bit confused about how this works (or ought to work) with asymmetric filters. Right now, I imagine it does weird things.

Better unit testing would expose this.